rfcs/011-resource-access-control/011-resource-access-control.md

```diff
@@ -0,0 +1,186 @@
+
+Each of those two gates is implemented by different services, the first is an in-house service that exposes through an
+HTTP interface if a specific user is granted access to a resource pool, referred from here on as the **Resource Access Control**
+service, with the data kept in a Postgres database.
+The second gate is implemented through [Kubernetes resource quotas](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+to define the resource pools and the [resource quota per PriorityClass](https://kubernetes.io/docs/concepts/policy/resource-quotas/#resource-quota-per-priorityclass)
+to associate user sessions to resource pools.
+
+![Basic architecture diagram](./overall-flow.png)
+
+### The Resource Access Control Service
+
+The service exposes an HTTP API that the administrators interact with to create/update/delete resource pools and
+add/remove users to resource pools.
+
+A prototype of the API exposed by the service can be seen in the attached file: [api-prototype](./api-prototype.yaml)
+
+From the above API, it is clear that the service needs to be able to create `ResourceQuota` and `PriorityClass` resources
+in the Kubernetes cluster and create the association between users and resource pools in the Postgres database.
+
+A non-essential feature would include a dedicated web page where the Renkulab administrator can log in and interact with
+the above API through a user interface.  The interface would also surface errors reported from the API which verifies if
+the passed request is possible.
+
+### Database structure
+
+The database serving as the backend to the Resource Access Control service will have four tables:
+
+```
+Quota:
+ - CPU
+ - memory
+ - GPU
+ - storage
+
+Resource class:
+ - name
+ - CPU
+ - memory
+ - GPU
+ - storage
+
+Resource pool:
+ - name
+ - classes (list of references to resource class)
+ - quota (reference to Quota)
+
+User:
+ - Keycloak ID
+ - pools (list of references to resource pools the user has access to)
+```
+
+### Advanced resource pools management
+
+In some cases, the administrator will need to dedicate specific Kubernetes nodes to resource pools, for instance when
+some user groups are allowed to use GPUs or when the autoscaling profiles for the nodes are different.
+
+In those cases, we rely on the `affinity` and `taint/toleration` mechanisms from Kubernetes. For each resource pool, the
+administrator will take care of:
+ - (optional) Setting a label on the nodes dedicated to the resource pool
+ - Setting a taint on the nodes dedicated to the resource pool
+ - Specify toleration and (optionally) a label when creating/updating the resource pool associated with the nodes
+
+The above will result in three different modes of operation:
+
+1. No node taints/tolerations or affinities
+  - sessions can be scheduled on any default or untainted session node (subject to resource availability and node sizes)
+  - resource quotas still apply (based on priority class)
```

Comment:
what is a priority class? I didn't think we were defining them?

Reply:
Priority class is a k8s resource. It allows us have quotas for resources that are assigned by k8s without having 1 k8s namespace per quota.